### PR TITLE
[clang] Lifetime of locals must end before musttail call

### PR DESCRIPTION
**Pull request by ostannard**
The lifetimes of local variables and function parameters must end before the call to a [[clang::musttail]] function, instead of before the return, because we will not have a stack frame to hold them when doing the call.

This documents this limitation, and adds diagnostics to warn about some code which is invalid because of it.

### CODE REVIEW

**Review by hokein — COMMENTED** [review]
I think it warrants a mention in the release notes.

File: clang/test/SemaCXX/attr-musttail.cpp [hunk 1]
```diff
@@ -267,3 +267,30 @@ namespace ns {}
...
   [[clang::musttail]] return ns; // expected-error {{unexpected namespace name 'ns': expected expression}}
 }
+
+// Test diagnostics for lifetimes of local variables, which end earlier for a
+// musttail call than for a nowmal one.
+
+void TakesIntAndPtr(int, int *);
+void PassAddressOfLocal(int a, int *b) {
+  int c;
+  [[clang::musttail]] return TakesIntAndPtr(0, &c); // expected-warning {{address of stack memory associated with local variable 'c' passed to musttail function}}
```
Comment (hokein) [inline clang/test/SemaCXX/attr-musttail.cpp:277]: can you add a test where a function call is used in the argument? something like

```
int foo();
void test(int a, const int &b) {
  int c;
  [[clang::musttail]] return test(0, foo()); 
}
```

File: clang/include/clang/Basic/AttrDocs.td [hunk 2]
```diff
@@ -637,6 +637,11 @@ return value must be trivially destructible. The calling convention of the
 caller and callee must match, and they must not be variadic functions or have
 old style K&R C function declarations.
 
+The lifetimes of all local variables and function parameters end immediately
+before the call to the function. This means that it is undefined behaviour to
+pass a pointer or reference to a local variable to the called function, which
+is not the case without the attribute.
```
Comment (hokein) [inline clang/include/clang/Basic/AttrDocs.td:643]: nit: maybe mention that clang will give a diagnostic on this case.

**Review by hokein — APPROVED** [review]
Thanks, it looks good, just a few nits.

File: clang/test/SemaCXX/attr-musttail.cpp (hunk 1 above)
Comment (ostannard) [reply clang/test/SemaCXX/attr-musttail.cpp:277]: Done, and there was a different diagnostic message used for temporaries, so added a better wording for that.

**Review by hokein — APPROVED** [review]
Thanks, still looks good.

**Comment by stefson** [comment]
@ostannard can you please backport this patch into the 19.x branch, and also other relevant changes to the musttail issue? clang-19.1.0 was released very broken on arm due to this.

**Pull request merged**